### DISCUSSION TITLE
brew install --build-bottle will not let me rebuild a bottle i've already built.

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#1850] [Tap maintenance and Homebrew development]
for my specific use case, i'm working with bottles and formulae related to freecad and the homebrew-freecad tap.

i like to edit formula files locally. i have the below path on my local file system where i can edit the files locally.

```
/opt/code/github/public/forks/homebrew-freecad
```

i then can cd into temporary directory such as,

```
cd ~/ltmp
```

and run,

```
brew install --build-bottle $ghforks/homebrew-freecad/Formula/boost-python3@1.75.0.rb --verbose
```

and i get the below output

```
Error: Failed to load cask: /opt/code/github/public/forks/homebrew-freecad/Formula/boost-python3@1.75.0.rb
Cask 'boost-python3@1.75.0' is unreadable: wrong constant name #<Class:0x00007facf84ceff8>
Warning: Treating /opt/code/github/public/forks/homebrew-freecad/Formula/boost-python3@1.75.0.rb as a formula.
Warning: boost-python3@1.75.0 1.75.0 is already installed and up-to-date.
To reinstall 1.75.0, run:
  brew reinstall boost-python3@1.75.0
```

i could obviously remove the installed formula using `brew remove` but i'm hoping there's a more streamlined way to just rebuild the bottle without having to remove it first.

and i want to specify a local file on my system because i'm editing these file before pushing to a github repo that is a tap.

**Comment by SMillerDev** [comment]
Reading the manpage I'd say https://docs.brew.sh/Manpage#install-options-formulacask- `--force` is what you're looking for.

**Reply by ipatch** [reply]
unfortunately `--force` does **not** run through the install process again, rather the below happens.

**ex**

> any formula from the freecad tap can be used in this example

```
echo "initial install of formula"
brew install --build-bottle freecad/freecad/coin@4.0.0
brew postinstall freecad/freecad/coin@4.0.0
brew bottle --json freecad/freecad/coin@4.0.0
echo "the formula is installed and a bottle is placed in the $cwd the brew cmds were run from"

echo "make a change to the `coin@4.0.0.rb` formula file"
vim $brew_prefix/Homebrew/Library/Taps/freecad/homebrew-freecad/Formula/coin@4.0.0.rb

echo "changes made, and saved to local file on disk"

echo "rerun install and bottle process to experiment with changes made to formula file"

brew install --build-bottle freecad/freecad/coin@4.0.0
```

after that last install cmd is run, the below is outputted

```
Warning: freecad/freecad/coin@4.0.0 4.0.0_1 is already installed and up-to-date.
To reinstall 4.0.0_1, run:
  brew reinstall coin@4.0.0
```

rerun install cmd with `--force` flag

```
brew install --build-bottle freecad/freecad/coin@4.0.0 --verbose --force                                                                 
Warning: freecad/freecad/coin@4.0.0 4.0.0_1 is already installed, it's just not linked.
To link this version, run:
  brew link coin@4.0.0
```

attempt to link the formula

```
brew link coin@4.0.0

echo "--- output"

Warning: coin@4.0.0 is keg-only and must be linked with `--force`.

If you need to have this software first in your PATH instead consider running:
  echo 'fish_add_path /usr/local/opt/coin@4.0.0/bin' >> ~/.config/fish/config.fish
```

```
brew link coin@4.0.0 --force                                                                                                             
Linking /usr/local/Cellar/coin@4.0.0/4.0.0_1... 11 symlinks created.

If you need to have this software first in your PATH instead consider running:
  echo 'fish_add_path /usr/local/opt/coin@4.0.0/bin' >> ~/.config/fish/config.fish
```

reattempt to build bottle from **edited** formula file

```
brew install --build-bottle freecad/freecad/coin@4.0.0 --verbose --force                                                                 
Warning: freecad/freecad/coin@4.0.0 4.0.0_1 is already installed, it's just not linked.
To link this version, run:
  brew link coin@4.0.0
```

the last install cmd puts the `$USER` back at the link step

so from my empirical evidence, from what i can tell the only way to rebuild a formula that was built using the `--build-bottle` flag is to first remove the formula using `brew remove` and more than likely such formula will have dependencies, so for this use case the `--ignore-dependencies` will be required, after the formula is uninstalled the original `brew install --build-bottle` should run through the task of installing the formula and allow the `brew bottle --json` cmd/taks to follow after the install completes.

please let me know if there is a better workflow for all of this, so far this is what i'm doing to test updates to formula files i've updated / edited.

on a positive note, akin to what @carlocab stated, editing the files within the tap puts the edits under version control and the edits are applied to the `brew install` cmd, so if i make a syntax err it will show up when running a `brew install ...` thus allowing me to test my edits locally without having to push to remote repo, and yes i've made sure to put my edits in a personal fork/branch of the upstream repo/tap.

leaving all this here for a _variant_ of _future me_ 👴🏼 ⏳

**Reply by carlocab** [reply]
Good to know my comment helped. Git is a really powerful tool for working on and experimenting with code -- you should take advantage of it!

As for trying to build another bottle while a formula is already installed: I would've thought `brew reinstall` accepted `--build-bottle` as a flag. Sadly, it does not. However, I think we'd be open to a pull request that adds a `--build-bottle` flag to `brew reinstall`.

In the meantime, while you're experimenting with formula changes, you can do `brew reinstall --build-from-source`. It won't build a version of the formula you can do `brew bottle` with, but it will allow you to quickly test what your changes did for your source build.

When you're happy with where you've gotten with that, then you can do `brew uninstall --ignore-dependencies` in order to do `brew install --build-bottle`.

**Comment by carlocab** [comment]
> and i want to specify a local file on my system because i'm editing these file before pushing to a github repo that is a tap.

You should do it in your local copy of your Git repository. This is what remote Git repositories are for.
